libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,324 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * UnusedFunctionParameterPruner: Optimiser step that removes unused parameters from function
+ * definition.
+ */
+
+#include <libyul/optimiser/UnusedFunctionParameterPruner.h>
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/YulString.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <algorithm>
+#include <optional>
+#include <variant>
+
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::langutil;
+
+namespace
+{
+
+bool anyFalse(vector<bool> const& _mask)
+{
+	return any_of(_mask.begin(), _mask.end(), [](bool b){ return !b; });
+}
+
+template<typename T>
+vector<T> applyBooleanMask(vector<T> const& _vec, vector<bool> const& _mask)
+{
+	yulAssert(_vec.size() == _mask.size(), "");
+
+	vector<T> ret;
+
+	for (size_t i = 0; i < _mask.size(); ++i)
+		if (_mask[i])
+			ret.push_back(_vec[i]);
+
+	return ret;
+}
+
+/**
+ * Step 1 of UnusedFunctionParameterPruner:
+ *
+ * Find functions whose parameters (both arguments and return-parameters) are not used in its body.
+ */
+class FindFunctionsWithUnusedParameters
```

Comment:
I'm wondering whether we really need this to be a stateful class instead of just a function that is called on a FunctionDefinition